Source/Task/SpinLock.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+#include <algorithm>
+#include <atomic>
+#include <thread>
+
+//
+// SpinLock: A spinlock implementation based on std::atomic_flag that
+// prevents CPU starvation. SpinLock can be used as a RAII wrapper around
+// an external flag, or its static Lock API may be used to lock an
+// external flag.
+//
+class SpinLock
+{
+public:
+    SpinLock(_In_ std::atomic_flag& flag) : m_lock(flag)
+    {
+        Lock(m_lock);
+    }
+
+    ~SpinLock()
+    {
+        m_lock.clear(std::memory_order_release);
+    }
+
+    static void Lock(_In_ std::atomic_flag& flag)
+    {
+        unsigned int backoff = 1;
+        constexpr unsigned int maxBackoff = 1024;
+
+        while (flag.test_and_set(std::memory_order_acquire)) {
+            for (unsigned int i = 0; i < backoff; ++i) {
+                cpu_pause();
+            }
+            
+            // Exponential backoff with cap. If we are over the cap yield
+            // this thread.
+
+            backoff = backoff << 1;
+            if (backoff >= maxBackoff)
+            {
+                backoff = maxBackoff;
+                std::this_thread::yield();
+            }
+        }
+    }
+
+private:
+    std::atomic_flag& m_lock;
+
+    static inline void cpu_pause()
+    {
+#if defined(_M_IX86) || defined(_M_X64)
+        // x86/x64: Tells CPU we're spinning
+        // Reduces energy consumption
+        // Helps avoid memory order violations
+        // Maps to PAUSE instruction
+        _mm_pause();
```

Comment:
>_mm_pause(); [](http://example.com/codeflow?start=8&length=12)

One compiler issue I found is that this particular API is not supported on NDA console platform.

Reply:
need to add :
#if defined(_M_IX86) || defined(_M_X64)
#include <x86intrin.h>
#endif

Reply:
Is there another #ifdef we can check here? x86intrin.h doesn't exist on windows platforms.  In Windows OS code I see this pattern.  Will this work for the NDA platform?

#if defined( _MSC_VER )
    #include <intrin.h>
#elif defined( __GNUC__ )
    #include <x86intrin.h>
#endif

Reply:
Worked with Pushpadant offline and uploaded what we think will work for all platforms.